src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -375,7 +375,18 @@ private Task WriteHexInt32Async(int value)
                         // Proxied requests contain full URL
                         Debug.Assert(request.RequestUri.Scheme == Uri.UriSchemeHttp);
                         await WriteBytesAsync(s_httpSchemeAndDelimiter).ConfigureAwait(false);
-                        await WriteAsciiStringAsync(request.RequestUri.IdnHost).ConfigureAwait(false);
+
+                        // If the hostname is an IPv6 address, uri.IdnHost will return the address without enclosing [].
+                        // In this case, use uri.Host instead, which will correctly enclose with [].
+                        // Note we don't need punycode encoding if it's an IP address, so using uri.Host is fine.
+                        await WriteAsciiStringAsync(request.RequestUri.HostNameType == UriHostNameType.IPv6 ?
+                            request.RequestUri.Host : request.RequestUri.IdnHost).ConfigureAwait(false);
+
+                        if (!request.RequestUri.IsDefaultPort)
+                        {
+                            await WriteByteAsync((byte)':').ConfigureAwait(false);
+                            await WriteDecimalInt32Async(request.RequestUri.Port).ConfigureAwait(false);
```

Comment:
It seem costly to separate out the ":" character and the small number of decimal digits into separate async operations requiring Task's to be created and awaited, etc.  

In fact, this entire request line construction involves too many async Task-based operations.  It probably would be more efficient to buffer all these bytes into a single async Write call.

You don't need to block this PR for this but this seems like a worthwhile change to make for future perf gains.

cc: @geoffkizer

Reply:
We do want to revisit this, as there is overhead here we can likely do away with.  Note, though, that on the fast path here, we don't actually create any tasks; these methods end up returning an already completed task, such that nothing is allocated and the await sees IsCompleted==true and skips hooking up any callbacks.  That said, it's still extra code being executed that we could benefit from doing away with.

Reply:
We may also want to optimize [here](https://github.com/dotnet/corefx/blob/24889e293731f12b11f1205b641cd3e6cb3a3b90/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L304).